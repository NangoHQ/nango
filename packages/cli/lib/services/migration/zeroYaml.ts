import { spawn } from 'node:child_process';
import fs from 'node:fs';
import path from 'node:path';

import chalk from 'chalk';
import jscodeshift from 'jscodeshift';

import { printDebug } from '../../utils.js';
import { NANGO_VERSION } from '../../version.js';
import { compileAllFiles } from '../compile.service.js';
import { loadYamlAndGenerate } from '../model.service.js';
import verificationService from '../verification.service.js';

import type { Collection } from 'jscodeshift';

export async function migrateToZeroYaml({ fullPath, debug }: { fullPath: string; debug: boolean }) {
    await verificationService.necessaryFilesExist({ fullPath, autoConfirm: true, debug });
    const { success } = await compileAllFiles({ fullPath, debug });
    if (!success) {
        console.log(chalk.red('Failed to compile. Exiting'));
        process.exitCode = 1;
        return;
    }

    const parsed = loadYamlAndGenerate({ fullPath, debug });
    if (!parsed) {
        process.exitCode = 1;
        return;
    }

    await addPackageJson({ fullPath, debug });

    for (const integration of parsed.integrations) {
        for (const sync of integration.syncs) {
            const content = await getContent({ fullPath, integrationId: integration.providerConfigKey, scriptType: 'syncs', scriptName: sync.name });
            //             content = content.replace(
            //                 /import type.*models";/,
            //                 `import { createSync } from 'nango';
            // import { z } from 'zod';
            // import type { NangoSync } from 'nango';`
            //             );

            //             content.replace(
            //                 /export default async function fetchData.*/,
            //                 `export default createSync({
            //     description: '${sync.description.replaceAll("'", "\'")}',
            //     version: '${sync.version}',
            //     endpoints: [{ method: 'GET', path: '/example/github/issues', group: 'Issues' }],
            //     runs: '${sync.runs}',
            //     autoStart: ${sync.auto_start},
            //     syncType: '${sync.sync_type}',
            //     trackDeletes: ${sync.track_deletes},
            //     models: {
            //         GithubIssue: issueSchema
            //     },

            //     // Sync execution
            //     exec: async (nango) => {
            //       Function content
            // });`
            //             );

            const j = jscodeshift.withParser('ts');
            const root = j(content);

            // Add import { createSync } from 'nango'; if not present
            const hasCreateSyncImport =
                root
                    .find(j.ImportDeclaration)
                    .filter((path) => {
                        return (
                            path.node.source.value === 'nango' &&
                            Array.isArray(path.node.specifiers) &&
                            path.node.specifiers.some((spec) => spec.type === 'ImportSpecifier' && spec.imported && spec.imported.name === 'createSync')
                        );
                    })
                    .size() > 0;

            if (!hasCreateSyncImport) {
                const importDecl = j.importDeclaration([j.importSpecifier(j.identifier('createSync'))], j.literal('nango'));
                root.get().node.program.body.unshift(importDecl);
            }

            addZodImport(root, j);

            // Wrap default function
            root.find(j.ExportDefaultDeclaration).forEach((path) => {
                const func = path.node.declaration;
                if (func.type === 'FunctionDeclaration') {
                    // Create an object with exec property as the function expression
                    const execArrow = j.arrowFunctionExpression(func.params, func.body);
                    execArrow.async = true;
                    const execProp = j.objectProperty(j.identifier('exec'), execArrow);
                    const descriptionProp = j.objectProperty(j.identifier('description'), j.stringLiteral(sync.description));
                    const versionProp = j.objectProperty(j.identifier('version'), j.stringLiteral(sync.version || '0.0.1'));
                    const runsProp = j.objectProperty(j.identifier('runs'), j.stringLiteral(sync.runs));
                    const autoStartProp = j.objectProperty(j.identifier('autoStart'), j.booleanLiteral(sync.auto_start));
                    const syncTypeProp = j.objectProperty(j.identifier('syncType'), j.stringLiteral(sync.sync_type));
                    const trackDeletesProp = j.objectProperty(j.identifier('trackDeletes'), j.booleanLiteral(sync.track_deletes));
                    const props = [descriptionProp, versionProp, runsProp, autoStartProp, syncTypeProp, trackDeletesProp];
                    if (Array.isArray(sync.scopes) && sync.scopes.length > 0) {
                        const scopesProp = j.objectProperty(j.identifier('scopes'), j.arrayExpression(sync.scopes.map((s) => j.stringLiteral(s))));
                        props.push(scopesProp);
                    }
                    props.push(execProp);
                    const obj = j.objectExpression(props);
                    path.replace(j.exportDefaultDeclaration(j.callExpression(j.identifier('createSync'), [obj])));
                }
            });

            const transformed = root.toSource();

            // Append transformed code after line 55
            const targetFile = path.join(fullPath, integration.providerConfigKey, 'syncs', `${sync.name}.v2.ts`);
            await fs.promises.writeFile(targetFile, transformed);
        }
    }

    // await fs.promises.rm(path.join(fullPath, 'nango.yaml'));
    // await fs.promises.rm(path.join(fullPath, 'models.ts'));

    await runNpmInstall(fullPath);
}

async function getContent({
    fullPath,
    integrationId,
    scriptType,
    scriptName
}: {
    fullPath: string;
    integrationId: string;
    scriptType: 'syncs' | 'actions' | 'on-events';
    scriptName: string;
}): Promise<string> {
    const res = await fs.promises.readFile(path.join(fullPath, integrationId, scriptType, `${scriptName}.ts`));
    return res.toString();
}

/**
 * Adds a package.json file to the given directory if it doesn't exist
 * Otherwise, it updates the existing package.json file
 */
async function addPackageJson({ fullPath, debug }: { fullPath: string; debug: boolean }) {
    // Ensure package.json exists and has nango in devDependencies
    const packageJsonPath = path.join(fullPath, 'package.json');
    const examplePackageJsonPath = path.join(import.meta.dirname, '../../../example/package.json');
    let packageJsonExists = false;
    try {
        await fs.promises.access(packageJsonPath, fs.constants.F_OK);
        packageJsonExists = true;
    } catch (_err) {
        packageJsonExists = false;
    }

    if (!packageJsonExists) {
        printDebug('package.json does not exist', debug);
        await fs.promises.copyFile(examplePackageJsonPath, packageJsonPath);
    } else {
        printDebug('package.json exists, updating', debug);
        const pkgRaw = await fs.promises.readFile(packageJsonPath, 'utf-8');
        const pkg = JSON.parse(pkgRaw) as { devDependencies?: Record<string, string>; dependencies?: Record<string, string> };

        const examplePkgRaw = await fs.promises.readFile(examplePackageJsonPath, 'utf-8');
        const examplePkg = JSON.parse(examplePkgRaw);
        const zodVersion = (examplePkg.devDependencies && examplePkg.devDependencies['zod'])!;
        pkg.devDependencies = pkg.devDependencies || {};
        pkg.devDependencies['nango'] = NANGO_VERSION;
        pkg.devDependencies['zod'] = zodVersion;

        // Remove nango and zod from dependencies just in case they were added as prod
        if (pkg.dependencies?.['nango']) {
            delete pkg.dependencies['nango'];
        }
        if (pkg.dependencies?.['zod']) {
            delete pkg.dependencies['zod'];
        }
        await fs.promises.writeFile(packageJsonPath, JSON.stringify(pkg, null, 2));
    }
}

/**
 * Runs npm install in the given directory
 */
async function runNpmInstall(fullPath: string): Promise<void> {
    await new Promise((resolve, reject) => {
        const proc = spawn('npm', ['install', '--no-audit', '--no-fund', '--no-progress'], {
            cwd: fullPath,
            stdio: 'inherit',
            shell: true
        });
        proc.on('close', (code) => {
            if (code === 0) {
                resolve(undefined);
            } else {
                reject(new Error(`npm install failed with exit code ${code}`));
            }
        });
    });
}

/**
 * Adds an import for zod if it's not present
 */
function addZodImport(root: Collection, j: jscodeshift.JSCodeshift) {
    const hasZodImport =
        root
            .find(j.ImportDeclaration)
            .filter((path) => {
                return (
                    path.node.source.value === 'zod' &&
                    Array.isArray(path.node.specifiers) &&
                    path.node.specifiers.some((spec) => spec.type === 'ImportSpecifier' && spec.imported && spec.imported.name === 'z')
                );
            })
            .size() > 0;

    if (!hasZodImport) {
        const importDecl = j.importDeclaration([j.importSpecifier(j.identifier('z'))], j.literal('zod'));
        root.get().node.program.body.unshift(importDecl);
    }
}
