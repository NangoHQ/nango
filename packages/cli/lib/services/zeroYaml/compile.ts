import chalk from 'chalk';
import { build } from 'esbuild';
import { glob } from 'glob';
import ts from 'typescript';

import { Err, Ok } from '../../utils/result.js';

import type { Result } from '@nangohq/types';

const tsconfig: ts.CompilerOptions = {
    module: ts.ModuleKind.CommonJS,
    target: ts.ScriptTarget.ESNext,
    strict: true,
    esModuleInterop: true,
    skipLibCheck: true,
    forceConsistentCasingInFileNames: true,
    moduleResolution: ts.ModuleResolutionKind.NodeJs,
    allowUnusedLabels: false,
    allowUnreachableCode: false,
    exactOptionalPropertyTypes: true,
    noFallthroughCasesInSwitch: true,
    noImplicitOverride: true,
    noImplicitReturns: true,
    noPropertyAccessFromIndexSignature: true,
    noUncheckedIndexedAccess: true,
    noUnusedLocals: true,
    noUnusedParameters: true,
    declaration: false,
    sourceMap: true,
    composite: false,
    checkJs: false,
    noEmit: true
};

function typeCheck({ entryPoints }: { entryPoints: string[] }): Result<boolean> {
    const program = ts.createProgram({
        rootNames: entryPoints,
        options: {
            ...tsconfig
        }
    });

    const diagnostics = ts.getPreEmitDiagnostics(program);
    if (diagnostics.length <= 0) {
        return Ok(true);
    }

    diagnostics.forEach((diagnostic) => {
        const message = ts.flattenDiagnosticMessageText(diagnostic.messageText, '\n');
        if (diagnostic.file && diagnostic.start != null) {
            const { line, character } = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start);
            const fileName = diagnostic.file.fileName;
            console.error(chalk.red('err'), '-', `${chalk.blue(fileName)} ${chalk.yellow(`:${line + 1}:${character + 1}`)}`, `\r\n  ${message}\r\n`);
        } else {
            console.error(message);
        }
    });

    console.error(`Found ${diagnostics.length} error${diagnostics.length > 1 ? 's' : ''}`);
    return Err('errors');
}

export async function compileAll({ fullPath }: { fullPath: string }): Promise<Result<boolean>> {
    try {
        const entryPoints = await glob('**/*.ts', { cwd: fullPath, ignore: ['node_modules/**', 'dist/**', 'build/**'] });

        const typechecked = typeCheck({ entryPoints });
        if (typechecked.isErr()) {
            return typechecked;
        }

        await build({
            entryPoints: entryPoints,
            outdir: 'build',
            bundle: false,
            sourcemap: true,
            format: 'cjs',
            target: 'esnext',
            platform: 'node',
            outbase: '.',
            tsconfigRaw: {
                compilerOptions: {
                    module: 'commonjs',
                    target: 'esnext',
                    strict: true,
                    esModuleInterop: true,
                    skipLibCheck: true,
                    forceConsistentCasingInFileNames: true,
                    moduleResolution: 'node',
                    allowUnusedLabels: false,
                    allowUnreachableCode: false,
                    exactOptionalPropertyTypes: true,
                    noFallthroughCasesInSwitch: true,
                    noImplicitOverride: true,
                    noImplicitReturns: true,
                    noPropertyAccessFromIndexSignature: true,
                    noUncheckedIndexedAccess: true,
                    noUnusedLocals: true,
                    noUnusedParameters: true,
                    declaration: false,
                    sourceMap: true,
                    composite: false,
                    checkJs: false
                }
            },
            logLevel: 'info'
        });
        return Ok(true);
    } catch (err) {
        console.error(err);
        return Err('failed_to_compile');
    }
}
