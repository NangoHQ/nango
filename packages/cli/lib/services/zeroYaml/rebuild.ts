// import { createRequire } from 'node:module';
import path from 'node:path';

import { Ok } from '../../utils/result.js';

import type { NangoYamlParsed, Result } from '@nangohq/types';

// const require = createRequire(import.meta.url);

export async function rebuildParsed({ fullPath }: { fullPath: string; debug: boolean }): Promise<Result<NangoYamlParsed>> {
    const content = await import(path.join(fullPath, 'build', 'index.cjs'));
    console.log(content);
    const parsed: NangoYamlParsed = { yamlVersion: 'v2', integrations: [], models: new Map() };

    return Ok(parsed);
}
