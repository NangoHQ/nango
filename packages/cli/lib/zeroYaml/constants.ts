import ts from 'typescript';

export const npmPackageRegex = /^[^./\s]/;
export const importRegex = /^import ['"](?<path>\.\/[^'"]+)['"];/gm;

export const tsconfig: ts.CompilerOptions = {
    module: ts.ModuleKind.CommonJS,
    target: ts.ScriptTarget.ESNext,
    strict: true,
    esModuleInterop: true,
    skipLibCheck: true,
    forceConsistentCasingInFileNames: true,
    moduleResolution: ts.ModuleResolutionKind.NodeJs,
    allowUnusedLabels: false,
    allowUnreachableCode: false,
    exactOptionalPropertyTypes: true,
    noFallthroughCasesInSwitch: true,
    noImplicitOverride: true,
    noImplicitReturns: true,
    noPropertyAccessFromIndexSignature: true,
    noUncheckedIndexedAccess: true,
    noUnusedLocals: true,
    noUnusedParameters: true,
    declaration: false,
    sourceMap: true,
    composite: false,
    checkJs: false,
    noEmit: true
};
export const tsconfigString: Record<string, any> = {
    ...tsconfig,
    module: 'commonjs',
    target: 'esnext',
    importsNotUsedAsValues: 'remove',
    jsx: 'react',
    moduleResolution: 'node16'
};

export type BabelErrorType = 'nango_unsupported_export' | 'nango_invalid_function_param' | 'nango_invalid_default_export' | 'nango_invalid_export_constant';
export class BabelError extends Error {
    type;
    constructor(type: BabelErrorType) {
        super(type);
        this.type = type;
    }
}

const defaultMsg = 'Invalid default export: should be createAction(), createSync() or createOnEvent()';
export const customErrors: Record<BabelErrorType, string> = {
    nango_unsupported_export: defaultMsg,
    nango_invalid_function_param: 'Invalid function parameter, should be an object',
    nango_invalid_default_export: defaultMsg,
    nango_invalid_export_constant: defaultMsg
};
export class CompileError extends Error {
    type: BabelErrorType | 'failed_to_build_unknown';
    msg: string;
    filePath?: string;
    constructor(type: BabelErrorType | 'failed_to_build_unknown', msg: string) {
        super(msg);
        this.type = type;
        this.msg = msg;
    }
}
