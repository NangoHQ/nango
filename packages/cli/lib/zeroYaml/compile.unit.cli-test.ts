import { exec as execCb } from 'node:child_process';
import fs from 'node:fs';
import path from 'node:path';
import { promisify } from 'node:util';

import { describe, expect, it } from 'vitest';

import { bundleFile, compileAll } from './compile.js';
import { copyDirectoryAndContents, fixturesPath, getTestDirectory } from '../tests/helpers.js';

const exec = promisify(execCb);

describe('bundleFile', () => {
    it('should bundle a sync with a constant export', async () => {
        const result = await bundleFile({ entryPoint: path.join(fixturesPath, 'zero/valid/github/syncs/fetchIssues.ts') });
        const value = result.unwrap();
        expect(value).toMatchSnapshot();
    });
    it('should bundle an action with a default export', async () => {
        const result = await bundleFile({ entryPoint: path.join(fixturesPath, 'zero/valid/github/actions/createIssue.ts') });
        const value = result.unwrap();
        expect(value).toMatchSnapshot();
    });
});

describe('compileAll', () => {
    it('should compile a minimal integration', async () => {
        const dir = await getTestDirectory('zero_valid');
        console.log('compiling to ', dir);
        await copyDirectoryAndContents(path.join(fixturesPath, 'zero/valid'), dir);

        const pkg = { name: 'test', dependencies: { nango: `file:${path.resolve(path.join(fixturesPath, '..'))}`, zod: '3.24.2' } };

        await fs.promises.writeFile(path.join(dir, 'package.json'), JSON.stringify(pkg, null, 2));
        await exec('npm i', { cwd: dir });
        const result = await compileAll({ fullPath: dir, debug: false });
        result.unwrap();
        expect(result.isOk()).toBe(true);
    });
});
