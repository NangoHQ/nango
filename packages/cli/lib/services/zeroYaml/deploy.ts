import fs from 'node:fs';
import path from 'node:path';

import { isAxiosError } from 'axios';
import chalk from 'chalk';

import { rebuildParsed } from './rebuild.js';
import { Err, Ok } from '../../utils/result.js';
import { enrichHeaders, hostport, http, parseSecretKey, printDebug } from '../../utils.js';

import type { DeployOptions } from '../../types.js';
import type {
    CLIDeployFlowConfig,
    NangoConfigMetadata,
    NangoYamlParsed,
    OnEventScriptsByProvider,
    OnEventType,
    PostDeploy,
    Result,
    ScriptFileType
} from '@nangohq/types';
import type { AxiosResponse } from 'axios';
import type { JSONSchema7 } from 'json-schema';

export async function deploy({
    fullPath,
    options,
    environmentName
}: {
    fullPath: string;
    options: DeployOptions;
    environmentName: string;
}): Promise<Result<boolean>> {
    const { version, sync: optionalSyncName, action: optionalActionName, debug } = options;

    const parsed = await rebuildParsed({ fullPath, debug });
    if (parsed.isErr()) {
        return Err(parsed.error);
    }

    await parseSecretKey(environmentName, debug);

    const postData = await createPackage({ parsed: parsed.value, fullPath, debug, version, optionalSyncName, optionalActionName });
    if (!postData) {
        return Err('no_data');
    }

    const nangoYamlBody = '';

    const url = hostport + `/sync/deploy`;

    const bodyDeploy: PostDeploy['Body'] = { ...postData, reconcile: true, debug, nangoYamlBody };

    return await postPackage(url, bodyDeploy);
}

async function createPackage({
    parsed,
    fullPath,
    debug,
    version = '',
    optionalSyncName,
    optionalActionName
}: {
    parsed: NangoYamlParsed;
    fullPath: string;
    debug: boolean;
    version?: string | undefined;
    optionalSyncName?: string | undefined;
    optionalActionName?: string | undefined;
}): Promise<{
    flowConfigs: CLIDeployFlowConfig[];
    onEventScriptsByProvider: OnEventScriptsByProvider[] | undefined;
    jsonSchema: JSONSchema7 | undefined;
} | null> {
    const postData: CLIDeployFlowConfig[] = [];
    const onEventScriptsByProvider: OnEventScriptsByProvider[] | undefined = optionalActionName || optionalSyncName ? undefined : []; // only load on-event scripts if we're not deploying a single sync or action

    for (const integration of parsed.integrations) {
        const { providerConfigKey, onEventScripts } = integration;

        if (onEventScriptsByProvider) {
            const scripts: OnEventScriptsByProvider['scripts'] = [];
            for (const event of Object.keys(onEventScripts) as OnEventType[]) {
                for (const scriptName of onEventScripts[event]) {
                    const files = await loadScriptFiles({ scriptName, providerConfigKey, fullPath, type: 'on-events' });
                    if (!files) {
                        console.log(chalk.red(`No script files found for "${scriptName}"`));
                        return null;
                    }
                    scripts.push({ name: scriptName, fileBody: files, event });
                }
            }

            if (scripts.length > 0) {
                onEventScriptsByProvider.push({ providerConfigKey, scripts });
            }
        }

        if (!optionalActionName) {
            for (const sync of integration.syncs) {
                if (optionalSyncName && optionalSyncName !== sync.name) {
                    continue;
                }

                const metadata: NangoConfigMetadata = {};
                if (sync.description) {
                    metadata['description'] = sync.description;
                }
                if (sync.scopes) {
                    metadata['scopes'] = sync.scopes;
                }

                const files = await loadScriptFiles({ scriptName: sync.name, providerConfigKey, fullPath, type: 'syncs' });
                if (!files) {
                    console.log(chalk.red(`No script files found for "${sync.name}"`));
                    return null;
                }
                if (debug) {
                    printDebug(`Scripts files found for ${sync.name}`);
                }

                const body: CLIDeployFlowConfig = {
                    syncName: sync.name,
                    providerConfigKey,
                    models: sync.output || [],
                    version: version || sync.version,
                    runs: sync.runs,
                    track_deletes: sync.track_deletes,
                    auto_start: sync.auto_start,
                    attributes: {},
                    metadata: metadata,
                    input: sync.input || undefined,
                    sync_type: sync.sync_type,
                    type: sync.type,
                    fileBody: files,
                    model_schema: sync.usedModels.map((name) => parsed.models.get(name)!),
                    endpoints: sync.endpoints,
                    webhookSubscriptions: sync.webhookSubscriptions
                };

                postData.push(body);
            }
        }

        if (!optionalSyncName) {
            for (const action of integration.actions) {
                if (optionalActionName && optionalActionName !== action.name) {
                    continue;
                }

                const metadata = {} as NangoConfigMetadata;
                if (action.description) {
                    metadata['description'] = action.description;
                }
                if (action.scopes) {
                    metadata['scopes'] = action.scopes;
                }

                const files = await loadScriptFiles({ scriptName: action.name, providerConfigKey, fullPath, type: 'actions' });
                if (!files) {
                    console.log(chalk.red(`No script files found for "${action.name}"`));
                    return null;
                }
                if (debug) {
                    printDebug(`Scripts files found for "${action.name}"`);
                }

                const body: CLIDeployFlowConfig = {
                    syncName: action.name,
                    providerConfigKey,
                    models: action.output || [],
                    version: version || action.version,
                    runs: '',
                    metadata: metadata,
                    input: action.input || undefined,
                    type: action.type,
                    fileBody: files,
                    model_schema: action.usedModels.map((name) => parsed.models.get(name)!),
                    endpoints: action.endpoint ? [action.endpoint] : [],
                    track_deletes: false
                };

                postData.push(body);
            }
        }
    }

    // const jsonSchema = loadSchemaJson({ fullPath });
    // if (!jsonSchema) {
    //     return null;
    // }

    // console.log(postData);

    return {
        flowConfigs: postData,
        onEventScriptsByProvider,
        jsonSchema: undefined
    };
}

async function loadScriptFiles({
    fullPath,
    scriptName,
    providerConfigKey,
    type
}: {
    fullPath: string;
    scriptName: string;
    providerConfigKey: string;
    type: ScriptFileType;
}): Promise<{ js: string; ts: string } | null> {
    const js = await loadScriptJsFile({ fullPath, scriptName, providerConfigKey, type });
    if (!js) {
        return null;
    }

    const ts = await loadScriptTsFile({ fullPath, scriptName, providerConfigKey, type });
    if (!ts) {
        return null;
    }

    return { js, ts };
}

async function loadScriptJsFile({
    scriptName,
    providerConfigKey,
    fullPath,
    type
}: {
    scriptName: string;
    type: ScriptFileType;
    providerConfigKey: string;
    fullPath: string;
}): Promise<string | null> {
    const filePath = path.join(fullPath, 'dist', providerConfigKey, type, `${scriptName}.cjs`);

    try {
        const content = await fs.promises.readFile(filePath, 'utf8');
        return content;
    } catch (err) {
        console.error(chalk.red(`Error loading file ${filePath}`), err instanceof Error ? err.message : err);
        return null;
    }
}

async function loadScriptTsFile({
    fullPath,
    scriptName,
    providerConfigKey,
    type
}: {
    fullPath: string;
    scriptName: string;
    providerConfigKey: string;
    type: ScriptFileType;
}): Promise<string | null> {
    const filePath = path.join(fullPath, providerConfigKey, type, `${scriptName}.ts`);

    try {
        const content = await fs.promises.readFile(filePath, 'utf8');
        return content;
    } catch (err) {
        console.error(chalk.red(`Error loading file ${filePath}`), err instanceof Error ? err.message : err);
        return null;
    }
}

async function postPackage(url: string, body: PostDeploy['Body']): Promise<Result<boolean>> {
    console.log('Deploying', body.flowConfigs.length, 'scripts', ', to', url);

    try {
        const res = await http.post<any, AxiosResponse<PostDeploy['Success']>>(url, body, { headers: enrichHeaders() });
        if (res.data.length === 0) {
            console.log(chalk.green(`Successfully removed the syncs/actions.`));
        } else {
            const nameAndVersions = res.data.map((result) => `${result.name}@v${result.version}`);
            console.log(chalk.green(`Successfully deployed the scripts: ${nameAndVersions.join(', ')}!`));
        }
        return Ok(true);
    } catch (err) {
        const errorMessage = isAxiosError(err) ? JSON.stringify(err.response?.data, null, 2) : JSON.stringify(err, ['message', 'name', 'stack'], 2);
        console.log(chalk.red(`Error deploying the scripts with the following error: ${errorMessage}`));
        return Err('err');
    }
}
