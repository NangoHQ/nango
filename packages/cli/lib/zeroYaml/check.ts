import fs, { existsSync, readFileSync } from 'fs';
import path from 'path';

import chalk from 'chalk';
import ora from 'ora';
import semver from 'semver';

import { exampleFolder } from './constants.js';
import { runPackageManagerInstall } from '../migrations/toZeroYaml.js';
import { Err, Ok } from '../utils/result.js';
import { printDebug } from '../utils.js';
import { NANGO_VERSION } from '../version.js';

import type { Result } from '@nangohq/types';
import type { PackageJson } from 'type-fest';

export async function checkAndSyncPackageJson({ fullPath, debug }: { fullPath: string; debug: boolean }): Promise<Result<{ updated: boolean }>> {
    printDebug('Checking and syncing package.json', debug);

    let updated = false;
    let newPkg: PackageJson | undefined;
    const packageJsonPath = path.join(fullPath, 'package.json');
    try {
        const examplePackageJsonPath = path.join(exampleFolder, 'package.json');
        const examplePkgRaw = await fs.promises.readFile(examplePackageJsonPath, 'utf-8');
        const examplePkg = JSON.parse(examplePkgRaw) as PackageJson;

        if (!existsSync(packageJsonPath)) {
            updated = true;
            newPkg = examplePkg;
        } else {
            const packageJson = JSON.parse(readFileSync(packageJsonPath, 'utf-8')) as PackageJson;
            if (!packageJson.devDependencies) {
                packageJson.devDependencies = {};
                updated = true;
            }
            const userZodVersion = packageJson.devDependencies['zod'];
            const nangoZodVersion = examplePkg.devDependencies!['zod']!;

            if (!userZodVersion) {
                // No zod version installed, add it
                updated = true;
                packageJson.devDependencies['zod'] = nangoZodVersion;
            } else {
                // Check if the major version matches
                const userZodMajor = semver.major(semver.coerce(userZodVersion) || '0.0.0');
                const nangoZodMajor = semver.major(semver.coerce(nangoZodVersion) || '0.0.0');

                if (userZodMajor !== nangoZodMajor) {
                    // Different major version - force update
                    updated = true;
                    packageJson.devDependencies['zod'] = nangoZodVersion;
                } else if (userZodVersion !== nangoZodVersion) {
                    // Same major version, different minor/patch - warn but don't update
                    console.log(chalk.yellow(`⚠️  You are using zod version ${userZodVersion}, while Nango internally uses ${nangoZodVersion}.`));
                    console.log(
                        chalk.yellow(`   You may encounter compatibility issues. If you experience any problems, please update to ${nangoZodVersion}.`)
                    );
                }
            }
            if (!packageJson.devDependencies['nango'] || packageJson.devDependencies['nango'] !== NANGO_VERSION) {
                updated = true;
                packageJson.devDependencies['nango'] = NANGO_VERSION;
            }
            newPkg = packageJson;
        }
    } catch (err) {
        return Err(new Error(`Error checking package.json`, { cause: err }));
    }

    if (updated && newPkg) {
        console.log(chalk.yellow('Your dependencies are out of date. Updating...'));
        const spinner = ora({ text: 'Updating package.json' }).start();
        await fs.promises.writeFile(packageJsonPath, JSON.stringify(newPkg, null, 2));
        await runPackageManagerInstall(fullPath);
        spinner.succeed();
    } else {
        printDebug('Package.json is up to date', debug);
    }

    return Ok({ updated });
}
