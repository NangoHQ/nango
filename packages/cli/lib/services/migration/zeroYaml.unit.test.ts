import { describe, expect, it } from 'vitest';

import * as zeroYaml from './zeroYaml.js';

import type { NangoModel } from '@nangohq/types';

describe('transformSync', () => {
    it('should create a .v2.ts file with createSync and correct properties', () => {
        const ts = `export default async function fetchData(nango) {\n  await nango.log('hello');\n}`;
        const result = zeroYaml.transformSync({
            content: ts,
            sync: {
                type: 'sync',
                name: 'test',
                description: 'Test sync',
                version: '1.2.3',
                runs: 'every hour',
                auto_start: true,
                sync_type: 'full',
                track_deletes: false,
                scopes: ['repo', 'user'],
                endpoints: [{ method: 'GET', path: 'top', group: 'foobar' }],
                usedModels: ['Metadata', 'Model', 'ModelDynamic'],
                webhookSubscriptions: ['*'],
                input: 'Metadata',
                output: ['Model', 'ModelDynamic']
            },
            models: new Map<string, NangoModel>([
                [
                    'Model',
                    {
                        name: 'Model',
                        fields: [
                            { name: 'id', value: 'string', tsType: true },
                            { name: 'arrTsType', value: 'string', tsType: true, array: true },
                            { name: 'null', value: 'null', tsType: true },
                            { name: 'bool', optional: false, tsType: true, value: 'boolean' },
                            { name: 'any', tsType: true, value: 'any' },
                            { name: 'opt', value: 'number', tsType: true, optional: true },
                            { name: 'ref', value: 'Foo', tsType: false, model: true, optional: true },
                            {
                                name: 'union',
                                union: true,
                                value: [
                                    { name: '0', value: 'literal1' },
                                    { name: '1', value: 'literal2' }
                                ],
                                tsType: false
                            },
                            {
                                name: 'array',
                                array: true,
                                value: [
                                    { name: '0', value: 'arr1' },
                                    { name: '1', value: 'arr2' }
                                ],
                                tsType: false
                            },
                            { name: 'obj', value: [{ name: 'nes', value: 'ted' }] },
                            {
                                name: 'dynamicObj',
                                optional: false,
                                value: [{ dynamic: true, name: '__string', optional: false, tsType: true, value: 'date' }]
                            }
                        ]
                    }
                ],
                [
                    'ModelDynamic',
                    {
                        name: 'ModelDynamic',
                        fields: [
                            { dynamic: true, name: '__string', optional: false, tsType: true, value: 'string' },
                            { name: 'id', optional: false, tsType: true, value: 'string' }
                        ]
                    }
                ],
                [
                    'Metadata',
                    {
                        name: 'Metadata',
                        fields: [{ name: 'foo', value: 'bar' }]
                    }
                ]
            ])
        });
        expect(result).toMatchSnapshot();
    });
});
