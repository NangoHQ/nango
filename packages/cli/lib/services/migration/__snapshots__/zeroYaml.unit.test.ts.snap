// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformSync > should create a .v2.ts file with createSync and correct properties 1`] = `
"import { z } from "zod";
import { createSync } from "nango";

export default createSync({
  description: "Test sync",
  version: "1.2.3",
  runs: "every hour",
  autoStart: true,
  syncType: "full",
  trackDeletes: false,

  endpoints: [{
    method: "GET",
    path: "top",
    group: "foobar"
  }],

  webhookSubscriptions: ["*"],
  scopes: ["repo", "user"],

  models: {
    Model: z.object({
      id: z.string(),
      arrTsType: z.string().array(),
      null: z.nullable(),
      bool: z.boolean(),
      any: z.any(),
      opt: z.number().optional(),
      ref: z.literal("Foo").optional(),
      union: z.union([z.literal("literal1"), z.literal("literal2")]),

      array: z.object({
        0: z.literal("arr1"),
        1: z.literal("arr2")
      }).array(),

      obj: z.object({
        nes: z.literal("ted")
      }),

      dynamicObj: z.record(z.date()).and(z.object({}))
    }),

    ModelDynamic: z.record(z.string()).and(z.object({
      id: z.string()
    }))
  },

  metadata: {
    Metadata: z.object({
      foo: z.literal("bar")
    })
  },

  exec: async nango => {
    await nango.log('hello');
  }
});"
`;
